### PR TITLE
Improve get_labels

### PR DESCRIPTION
**Pull request by rimms**
Fixes https://github.com/jubatus/jubatus_core/issues/204

* ``liner_classifier``: Returns labels and number of ``train``
* ``nearest_neighbor_classifier``: Returns labels and number of id

This patch is based on #271.

Related patches are followings.
- [x] core (This PR)
- [x] server (https://github.com/jubatus/jubatus/pull/1106)
- [x] python client (https://github.com/jubatus/jubatus-python-client/pull/67)
- [x] ruby client (https://github.com/jubatus/jubatus-ruby-client/pull/38)
- [x] java client (https://github.com/jubatus/jubatus-java-client/pull/36)
- [x] cpp client (https://github.com/jubatus/jubatus/pull/1106)
- [x] go client (Nothing to do)
- [x] website (https://github.com/jubatus/website/pull/260)
- [x] jubadump (https://github.com/jubatus/jubadump/pull/31)

### CODE REVIEW

File: jubatus/core/storage/labels.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+void labels::increment(const std::string& label) {
+  util::concurrent::scoped_wlock lock(mutex_);
+  diff_[label] += 1;
+}
+
+bool labels::erase(const std::string label) {
+  util::concurrent::scoped_wlock lock(mutex_);
+  bool result = diff_.erase(label);
+  result += master_.erase(label);
```
Comment (kmaehashi) [inline jubatus/core/storage/labels.cpp]: (personally) I don't prefer `bool + bool` op.
How about `result |= master_.erase(label);`

File: jubatus/core/storage/labels.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+namespace core {
+namespace storage {
+
+labels::labels() {
+}
+
+labels::~labels() {
+}
+
+bool labels::set(const std::string& label) {
```
Comment (kmaehashi) [inline jubatus/core/storage/labels.cpp]: `set` sounds unnatural, as this data structure is a **counter**.
I think `add` is better.

File: jubatus/core/storage/labels.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+#include "../common/version.hpp"
+#include "../framework/mixable_helper.hpp"
+
+namespace jubatus {
+namespace core {
+namespace storage {
+
+class labels {
+ public:
+  typedef classifier::labels_t data_t;
```
Comment (kmaehashi) [inline jubatus/core/storage/labels.hpp]: Libraries under `storage` should not depend on `classifier` module.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp [hunk 4]
```diff
@@ -100,12 +108,18 @@ void nearest_neighbor_classifier::train(
   }
   nearest_neighbor_engine_->set_row(id, fv);
   set_label(label);
+
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_[label] += 1;
```
Comment (kmaehashi) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: Can't we use the same data structure (`jubatus::core::storage::labels`) in both NN-classifier and linear classifier?
Seems like doing a same thing twice.

File: jubatus/core/storage/labels.cpp (hunk 1 above)
Comment (rimms) [inline jubatus/core/storage/labels.cpp]: OK 👍 
I will change this calculation to the above.

File: jubatus/core/storage/labels.cpp (hunk 2 above)
Comment (rimms) [inline jubatus/core/storage/labels.cpp]: OK, I will rename this method to ``add`` .

File: jubatus/core/storage/labels.hpp (hunk 3 above)
Comment (rimms) [inline jubatus/core/storage/labels.hpp]: I will remove this dependency.

File: jubatus/core/classifier/nearest_neighbor_classifier.cpp (hunk 4 above)
Comment (rimms) [inline jubatus/core/classifier/nearest_neighbor_classifier.cpp]: That’s right. I try to change this patch to use ``storage::labels`` in NN-classifier.

**Comment by rimms** [comment]
@kmaehashi Thanks for review.
I rebased this patch and fixed.

**Comment by rimms** [comment]
I rebased. And I caught-up #276 .

File: jubatus/core/classifier/cosine_similarity_classifier.cpp [hunk 5]
```diff
@@ -43,14 +42,14 @@ void cosine_similarity_classifier::classify_with_scores(
     util::concurrent::scoped_rlock lk(storage_mutex_);
     mixable_storage_->get_model()->calc_scores(fv, ids, k_);
   }
+
+  labels_t labels = labels_.get_model()->get();
```
Comment (kmaehashi) [inline jubatus/core/classifier/cosine_similarity_classifier.cpp]: It's better to include `classifier_type.hpp` to use `labels_t`.

File: jubatus/core/storage/labels.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      return false;
+    }
+  }
+
+  util::concurrent::scoped_wlock lock(mutex_);
+  diff_[label] = 0;
+  return true;
+}
+
+labels::data_t labels::get() const {
```
Comment (kmaehashi) [inline jubatus/core/storage/labels.cpp]: As this is a counter, `get()` -> `get_labels()` is better I think.

File: jubatus/core/storage/labels.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+void labels::decrement(const std::string& label) {
+  util::concurrent::scoped_wlock lock(mutex_);
+  diff_[label] -= 1;
+  if (diff_[label] <= 0) {
+    diff_.erase(label);
+  }
+}
+
+bool labels::erase(const std::string label) {
```
Comment (kmaehashi) [inline jubatus/core/storage/labels.cpp]: `std::string&`

**Comment by kmaehashi** [comment]
@rimms Added a few comments.

File: jubatus/core/storage/labels.cpp (hunk 6 above)
Comment (kmaehashi) [inline jubatus/core/storage/labels.cpp]: As `data_t` is an internal data structure type we should not expose it as an API.

Comment (rimms) [inline jubatus/core/storage/labels.cpp]: This method returns the pairs of label and counter.
I rename this to ``get_labels`` .

File: jubatus/core/classifier/cosine_similarity_classifier.cpp (hunk 5 above)
Comment (rimms) [inline jubatus/core/classifier/cosine_similarity_classifier.cpp]: OK 👍

File: jubatus/core/storage/labels.cpp (hunk 7 above)
Comment (rimms) [inline jubatus/core/storage/labels.cpp]: I will fixed. Thank you.

**Comment by rimms** [comment]
@kmaehashi Thanks for review! I fixed ! 🏃 💨

**Comment by kmaehashi** [comment]
:+1: Great! Thank you!

**Pull request merged**